File: gold/17298.py
import sys
input = sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))
stack = []
answer = [0]*n

for i in range(n):
    while stack and a[stack[-1]] < a[i]:
        answer[stack.pop()] = a[i]
    stack.append(i)

while stack:
    answer[stack.pop()] = -1

for a in answer:
    print(a, end = ' ')

File: gold/17298.cpp
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int n;
	cin >> n;

	int tmp;
	vector<int> answer(n,0);
	stack<pair<int,int>> s;
	for (int i = 0; i < n; i++) {
		cin >> tmp;
		while(!s.empty() && tmp > s.top().first){
			answer[s.top().second] = tmp;
			s.pop();
		}
		s.push(make_pair(tmp,i));
	}
	while(!s.empty()) {
		answer[s.top().second] = -1;
		s.pop();
	}

	for (int i = 0; i < n; i++)
		cout << answer[i] << " ";
}